src/diagnostics/build.rs

```diff
@@ -22,21 +22,59 @@ pub fn collect_build_diagnostics(
         diagnostics
     });
 
-    let root_document = workspace.documents_by_uri.values().find(|document| {
-        document.data.as_latex().map_or(false, |data| {
-            !document.uri.as_str().ends_with(".aux")
-                && data
-                    .extras
-                    .implicit_links
-                    .log
-                    .iter()
-                    .any(|u| u.as_ref() == build_log_uri)
-        })
-    })?;
+    let root_document_uri = Arc::clone(
+        &workspace
+            .documents_by_uri
+            .values()
+            .find(|document| {
+                document.data.as_latex().map_or(false, |data| {
+                    !document.uri.as_str().ends_with(".aux")
+                        && data
+                            .extras
+                            .implicit_links
+                            .log
+                            .iter()
+                            .any(|u| u.as_ref() == build_log_uri)
+                })
+            })?
+            .uri,
+    );
+
+    let build_log_uri = Arc::clone(&build_log_document.uri);
 
-    let base_path = PathBuf::from(root_document.uri.path());
+    let mut base_path = PathBuf::from(root_document_uri.path());
+    base_path.pop();
     for error in &build_log.errors {
-        let position = Position::new(error.line.unwrap_or(0), 0);
+        let full_path = base_path.join(&error.relative_path);
+
+        let doc = if error.line.is_some() && error.hint.is_some() {
+            workspace.load(full_path.clone()).unwrap_or(None)
```

Comment:
Nope, but that seemed to be the best solution.
It might be worth to expose a way to "fetch" a document, and return `None` if not present, I'll add that.

Reply:
I switched to your method and that avoided to make `workspace` mutable, which is nicer for the diff.